src/lib/storage/lz4_segment.cpp

```diff
@@ -120,27 +140,257 @@ std::vector<pmr_string> LZ4Segment<pmr_string>::decompress() const {
+     */
+    auto cached_block = std::vector<char>{};
+    if (use_caching && *previous_block_index != start_block) {
+      cached_block = std::vector<char>{previous_block};
+    }
+
+    for (size_t block_index = start_block; block_index <= end_block; ++block_index) {
+      // Only decompress the current block if it's not cached.
+      if (!(use_caching && block_index == *previous_block_index)) {
+        _decompress_block_to_bytes(block_index, previous_block);
+      }
+
+      // Set the offset for the end of the string.
+      if (block_index == end_block) {
+        block_end_offset = end_offset % _block_size;
+      }
+
+      /**
+       * Extract the string from the current block via the offsets and append it to the result string stream.
+       * If the cached block is not the start block, the data is retrieved from the copy.
+       */
+      pmr_string partial_result;
+      if (use_caching && block_index == *previous_block_index && block_index != start_block) {
+        const auto start_offset_it = cached_block.cbegin() + block_start_offset;
+        const auto end_offset_it = cached_block.cbegin() + block_end_offset;
+        partial_result = pmr_string{start_offset_it, end_offset_it};
+      } else {
+        const auto start_offset_it = previous_block.cbegin() + block_start_offset;
+        const auto end_offset_it = previous_block.cbegin() + block_end_offset;
+        partial_result = pmr_string{start_offset_it, end_offset_it};
+      }
+      result_string << partial_result;
+
+      // After the first iteration, this is set to 0 since only the first block's start offset can't be equal to zero.
+      block_start_offset = 0u;
+    }
+    return std::make_pair(pmr_string{result_string.str()}, end_block);
+  }
+}
+
+template <typename T>
+T LZ4Segment<T>::decompress(const ChunkOffset& chunk_offset) const {
+  auto decompressed_block = std::vector<char>(_block_size);
+  return decompress(chunk_offset, std::nullopt, decompressed_block).first;
+}
+
 template <typename T>
 std::shared_ptr<BaseSegment> LZ4Segment<T>::copy_using_allocator(const PolymorphicAllocator<size_t>& alloc) const {
-  auto new_compressed_data = pmr_vector<char>{_compressed_data, alloc};
+  auto new_lz4_blocks = pmr_vector<pmr_vector<char>>{alloc};
+  for (const auto& block : _lz4_blocks) {
+    new_lz4_blocks.emplace_back(pmr_vector<char>{block, alloc});
+  }
   auto new_null_values = pmr_vector<bool>{_null_values, alloc};
+  auto new_dictionary = pmr_vector<char>{_dictionary, alloc};
 
-  if (_offsets.has_value()) {
-    auto new_offsets = pmr_vector<size_t>(*_offsets, alloc);
-    return std::allocate_shared<LZ4Segment>(alloc, std::move(new_compressed_data), std::move(new_null_values),
-                                            std::move(new_offsets), _decompressed_size);
+  if (_string_offsets.has_value()) {
+    auto new_string_offsets = pmr_vector<size_t>(*_string_offsets, alloc);
```

Comment:
As I mentioned I would prefer calling all container constructors with normal braces, but if you disagree you should use curly braces here as well for consistency.